### PR TITLE
fluid examples: audience

### PR DESCRIPTION
**Pull request by zhenmichael**
React demo for fluid audience 

UserId Selection
image

Initial client with one audience member: 
image

UserId Selection with ContainerId
image

Two clients with the same user id
image

Two clients with two audience members:
image

### CODE REVIEW

File: audience-demo/src/index.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import React from 'react';
+import ReactDOM from 'react-dom/client';
+import './index.css';
+import App from './App';
+
+const root = ReactDOM.createRoot(document.getElementById('root'));
+root.render(
```
Comment (sonalivdeshpande) [inline audience-demo/src/index.js:7]: add copyright headers--

/*!
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */

File: audience-demo/src/components/AudienceList/AudienceList.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
+import './AudienceList.css'
+
```
Comment (sonalivdeshpande) [inline audience-demo/src/components/AudienceList/AudienceList.js:2]: headers here as well

File: audience-demo/src/App.js [hunk 3]
```diff
@@ -0,0 +1,85 @@
+/* eslint-disable no-restricted-globals */
```
Comment (sonalivdeshpande) [inline audience-demo/src/App.js:1]: add `eslint` below headers

File: audience-demo/package.json [hunk 4]
```diff
@@ -0,0 +1,43 @@
+{
+  "name": "audience-demo",
```
Comment (sonalivdeshpande) [inline audience-demo/package.json:2]: "name": "@fluid-example/audience-demo",

File: audience-demo/README.md [hunk 5]
```diff
@@ -0,0 +1,37 @@
+## Available Scripts
+
+In the project directory, you can run:
+
+### `npm start`
+
+Runs the app in the development mode.\
+Open [http://localhost:3000](http://localhost:3000) to view it in your browser.
+
```
Comment (sonalivdeshpande) [inline audience-demo/README.md:9]: are you planning to have a follow-up change to update the readme for this example?

File: audience-demo/webpack.config.js [hunk 6]
```diff
@@ -0,0 +1,8 @@
+module.exports = {
+    resolve: {
+        fallback: { 
+            "url": require.resolve("url/"),
```
Comment (sonalivdeshpande) [inline audience-demo/webpack.config.js:4]: not sure why fallback is required here?

@RishhiB -- could this be something related to the polyfill issue: https://webpack.js.org/configuration/resolve/#resolvefallback

**Comment by sonalivdeshpande** [comment]
Add this example as part of the CI here: https://github.com/microsoft/FluidExamples/blob/main/build.yml

File: audience-demo/package.json [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    "react": "^18.2.0",
+    "react-dom": "^18.2.0",
+    "react-scripts": "5.0.1",
+    "url": "^0.11.0"
+  },
+  "scripts": {
+    "start": "react-scripts start",
+    "build": "react-scripts build",
+    "test": "react-scripts test",
+    "eject": "react-scripts eject",
```
Comment (sonalivdeshpande) [inline audience-demo/package.json:22]: add `test:report` in package.json to resolve CI failures

**Review by skylerjokiel — CHANGES_REQUESTED** [review]
Overall looking good.

Some notes in addition to the comments:

- We discussed extending the demo to support 3 buttons when the page loads instead of directly loading in. The buttons are "User A", "User B", "Random User". These will allow the user to see connections and see how the properties modify.
- Please update the ui so the content is more compact. This will make it easier to fit more users per page.
- Please add ; at the end of your break lines. I know you are writing in js and it doesn't enforce it but it's functionally better. @Sonali, maybe we can add linting? 
- use const for non-modified variables
- Please add more comment explaining what the code is doing. I know not all our examples have this but it makes the code more

File: audience-demo/package.json [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    "start:server": "npx @fluidframework/azure-local-service@latest",
+    "test:report": "echo No test for this example"
+  },
+  "eslintConfig": {
+    "extends": [
```
Comment (sonalivdeshpande) [inline audience-demo/package.json:27]: remove `eslintConfig` to resolve the eslint issues

File: audience-demo/src/SelectUser.js [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+            Enter Container Id:
+            <input type="text" id="containerIdInput" onChange={() => handleChange()}></input>
+          </div>
+          <nav>
+            <button type="submit" style={select}>
+              <Link to="/AudienceDisplay" state={{ userId: 'user1', containerId: containerId }}>User 1</Link>
+            </button>
+            <button type="submit" style={select}>
+              <Link to="/AudienceDisplay" state={{ userId: 'user2', containerId: containerId }}>User 2</Link>
+            </button>
```
Comment (skylerjokiel) [inline audience-demo/src/SelectUser.js:36]: You want to use useNavigate to do programmatic navigation using react-router. Right now the Link text does the navigation and not the button.

```js
    const navigate = useNavigate();
    const handleSubmit= (userId) => {
      navigate("/AudienceDisplay", { userId, containerId });
    }
    // ... 
    <button style={select} onClick={() => {handleSubmit("user1")}>User 1</button>
    <button style={select} onClick={() => {handleSubmit("user2")}>User 2</button>
    <button style={select} onClick={() => {handleSubmit("random")}>Random User</button>
```

**Review by skylerjokiel — CHANGES_REQUESTED** [review]
Getting there :)

File: audience-demo/src/App.js [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+
+  const handleSelectUser = useCallback((userId, containerId) => {
+    setDisplayAudience(true)
+    setUserId(userId);
+    setContainerId(containerId);
+  });
+
+  const handleContainerNotFound = useCallback(() => {
+    setDisplayAudience(false)
+  });
```
Comment (skylerjokiel) [inline audience-demo/src/App.js:23]: when you're using useCallback I believe you're supposed to connect state changes that should re-trigger 

```suggestion
  const handleSelectUser = useCallback((userId, containerId) => {
    setDisplayAudience(true)
    setUserId(userId);
    setContainerId(containerId);
  }, [displayAudience, userId, containerId]);

  const handleContainerNotFound = useCallback(() => {
    setDisplayAudience(false)
  }, [displayAudience]);
```

**Review by skylerjokiel — DISMISSED** [review]
This is looking great!

**Review by sonalivdeshpande — APPROVED** [review]
Looks good :shipit: 💯

**Pull request merged**